### PR TITLE
Braz-6 Update user endpoint

### PR DESCRIPTION
**Pull request by mhmBraz**
### This PR will:
- Update user endpoint

[BRAZ-6]: https://servioitalo.atlassian.net/browse/BRAZ-6?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: services/users/app/update_user.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	if err == nil && output.Id != id {
+		return nil, errors.New(exception.CodePermission)
+	}
+
+	if updateUser.Password != "" {
+		secret := os.Getenv("ENC_SECRET")
+		encryptionData, err := gu.encryption.Encrypt(ctx, secret, updateUser.Password)
+
+		fmt.Printf("%v", err)
```
Comment (mhmBraz) [inline services/users/app/update_user.go:64]: estava dando erro, pois não estava usando o err em outro lugar.

File: services/users/app/update_user.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		return nil, err
+	}
+
+	var userOutput UpdateUserOutput
+
+	err = gu.crudRepository.GetById(ctx, database.UsersCollection, id, &userOutput)
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (italoservio) [inline services/users/app/update_user.go:81]: Completamente desnecessário fazer duas queries no banco de dados. A função `UpdateById` tem que retornar o registro mais atual.

**Review by italoservio — CHANGES_REQUESTED** [review]

File: cmd/users/main.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ func main() {
 	usersV1.Get("/", userController.GetUserPaginated)
 	usersV1.Get("/:id", userController.GetUserById)
 	usersV1.Delete("/:id", userController.DeleteUserById)
+	usersV1.Patch("/:id", userController.UpdateUser)
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply cmd/users/main.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/cf68e0651835e260b0c1e97366432a06e579b060

File: packages/database/crud_repository.go [hunk 4]
```diff
@@ -227,3 +233,29 @@ func mapToBsonM[T any](m map[string]T) bson.M {
 	}
 	return doc
 }
+
+func (cr *CrudRepository) GetByEmail(
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply packages/database/crud_repository.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/2256bdddd988bc6a59f14c1f90a5dbaa0fe84425

File: packages/exception/http_exception.go [hunk 5]
```diff
@@ -63,6 +65,10 @@ func errorCodeToStruct(code string) *HTTPException {
 		response.StatusMessage = "Internal Server Error"
 		response.StatusCode = http.StatusInternalServerError
 		response.ErrorMessage = "An expected error occurred and the server could not deal with it"
+	case CodePermission:
+		response.StatusMessage = "Not have permission"
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply packages/exception/http_exception.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/e7787c9cca7780c0058da41e4435525e4c87d157

File: services/users/app/create_user_test.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ type TestingDependencies_TestCreateUser struct {
 	mockCrudRepository *mocks.MockCrudRepositoryInterface
 	mockUserRepository *mocks.MockUserRepositoryInterface
 	createUserImpl     *app.CreateUserImpl
+	updateUserImpl     *app.UpdateUserImpl
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/app/create_user_test.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/c3a226b862e96a33ef9ee71e34d2bb0de7263940

File: services/users/infra/http/user_controller.go [hunk 7]
```diff
@@ -72,6 +84,33 @@ func (uc *UserControllerImpl) CreateUser(c *fiber.Ctx) error {
...
+	id := c.Params("id")
+
+	updateUser, err := uc.updateUserImpl.Do(ctx, &app.UpdateUserInput{
+		FirstName: body.FirstName,
+		LastName:  body.LastName,
+		Email:     body.Email,
+		Type:      body.Type,
+		Password:  body.Password,
+		UpdatedAt: time.Now(),
+	}, id, app.UpdateUserOutput{})
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/3c1670c5b0fc28d198a943e5799e38c04ad99522

[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller.go]: pra o mock app.UpdateUserOutput{}

File: packages/database/crud_repository_test.go [hunk 8]
```diff
@@ -243,92 +243,104 @@ func TestCrudRepository_UpdateById(t *testing.T) {
...
+		docs := bson.D{
+			{Key: "_id", Value: mockId},
+			{Key: "foo", Value: "bar"},
+		}
+
+		nestedMt.AddMockResponses(mtest.CreateSuccessResponse(
+			primitive.E{Key: "ok", Value: 1},
+			primitive.E{Key: "value", Value: docs},
+		))
+		defer nestedMt.ClearMockResponses()
```
Comment (italoservio) [inline packages/database/crud_repository_test.go:324]: @mhmBraz mock do `findOneAndUpdate`

Comment (italoservio) [reply packages/database/crud_repository_test.go:324]: quando acabar o restante me fale que eu olho denovo

File: packages/exception/http_exception.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (italoservio) [reply packages/exception/http_exception.go]: Se alterou para "403 Forbidden" a StatusMessage deve ser "Forbidden" e não "Unauthorized"

[the remark this answers is not in the document]
Comment (mhmBraz) [reply packages/exception/http_exception.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/830768004d2dc70db805d3f68f9bc53d68f626ef

File: services/users/app/update_user.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+	"github.com/italoservio/braz_ecommerce/packages/database"
+	"github.com/italoservio/braz_ecommerce/packages/encryption"
+	"github.com/italoservio/braz_ecommerce/packages/exception"
+	"github.com/italoservio/braz_ecommerce/services/users/domain"
+	"github.com/italoservio/braz_ecommerce/services/users/infra/storage"
+)
+
+type UpdateUserInterface interface {
+	Do(ctx context.Context, updateUser *UpdateUserInput, id string, output UpdateUserOutput) (*UpdateUserOutput, error)
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/app/update_user.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/8db53118659d2a1eb5d2333082d8296e3437a05e

File: services/users/infra/http/user_controller.go [hunk 10]
```diff
@@ -2,7 +2,9 @@ package http
 
 import (
 	"errors"
+	"log/slog"
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller.go]: https://github.com/italoservio/braz_ecommerce/pull/18/commits/0963864dd2b8ab27d93248c2cbace4986b502748

File: services/users/infra/http/user_controller.go [hunk 11]
```diff
@@ -97,13 +136,6 @@ func (uc *UserControllerImpl) DeleteUserById(c *fiber.Ctx) error {
 	return c.SendStatus(http.StatusNoContent)
 }
 
-type GetUserPaginatedPayload struct {
-	Page    int      `query:"page" validate:"required,number,gt=0"`
-	PerPage int      `query:"per_page" validate:"required,number,gt=0,lte=100"`
-	Emails  []string `query:"email" validate:"omitempty,dive,email"`
-	Ids     []string `query:"id" validate:"omitempty,dive,mongodb"`
-}
-
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller.go]: Só tinha mudado de lugar, mas voltamos

File: services/users/app/update_user.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	Type      string    `json:"type" validate:"omitempty,min=1,max=100" bson:"type,omitempty"`
+	Password  string    `json:"password" validate:"omitempty,min=1,max=100" bson:"password,omitempty"`
+	UpdatedAt time.Time `json:"updated_at" validate:"omitempty" bson:"updated_at,omitempty"`
+}
+
+type UpdateUserOutput struct {
+	*domain.UserDatabaseNoPassword `bson:",inline"`
+}
+
+func (gu *UpdateUserImpl) Do(ctx context.Context, updateUser *UpdateUserInput, id string, output UpdateUserOutput) (*UpdateUserOutput, error) {
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/app/update_user.go]: ok

File: services/users/app/update_user.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+	err := gu.crudRepository.GetByEmail(ctx, database.UsersCollection, updateUser.Email, &output)
+
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/app/update_user.go]: ok

File: services/users/app/update_user.go (hunk 1 above)
Comment (mhmBraz) [reply services/users/app/update_user.go]: ok

**Review by italoservio — APPROVED** [review]

**Pull request merged**